### PR TITLE
@intern challenges/three

### PR DESCRIPTION
**Pull request by yashthipsay**

### CODE REVIEW

**Review by RCantu92 — CHANGES_REQUESTED** [review]
Hey Yash, I understand what you're attempting to do with this implementation of the bot. However, this won't work in production how you expect. That is because it won't be the same bot instance that is monitoring Ethereum as would be monitoring Optimism and Arbitrum. They would be three separate instances, and as such, wouldn't save state across one another. I.e. the array of alerts that you're updating and then fetching from would be isolated to each instance and not accessible by the other two instances.

So the approach is on the right path, but it does need further work. I would suggest you look more into the `getAlerts` method from the Forta SDK that you are attempting to implement your own version of. Remember, you would be able to use this in production. 😉

**Review by Vxatz — CHANGES_REQUESTED** [review]

File: Challenge-3/src/helper.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      address.toLowerCase() === OPT_ESCROW_ADDRESS.toLowerCase();
+    l2Alerts.alerts[0].metadata[isOptEscrow ? "optEscBal" : "abtEscBal"] =
+      balance;
+
+    return balance.toString();
+  }
+
+  public async getL1Alerts(chaindId: number): Promise<AlertsResponse> {
+    return await getAlerts({
+      alertId: "L2-SUPPLY",
```
Comment (Vxatz) [inline Challenge-3/src/helper.ts:60]: The alert ID is incorrect, as this is trying to fetch the L2 alerts while being on L2.

File: Challenge-3/src/findings.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    type: FindingType.Exploit,
+    protocol: `${network}`,
+    metadata: {
+      l1Escrow: l1Escrow,
+      l2Supply: l2Supply,
+    },
+  });
+};
+
+export const createL2Finding = (
```
Comment (Vxatz) [inline Challenge-3/src/findings.ts:22]:
```suggestion
export const createL1Finding = (
```

Should be "L1"

File: Challenge-3/src/agent.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    const helperInstance = new Helper(provider);
+
+    if (chainId === 1) {
+      // Concurrently fetch balances for specified escrow addresses
+      const [optBalance, abtBalance] = await Promise.all([
+        helperInstance.getL1Balance(OPT_ESCROW_ADDRESS, blockEvent.blockNumber),
+        helperInstance.getL1Balance(ABT_ESCROW_ADDRESS, blockEvent.blockNumber),
+      ]);
+
+      findings.push(createL2Finding(optBalance, abtBalance));
```
Comment (Vxatz) [inline Challenge-3/src/agent.ts:38]: There's one thing that I've missed yesterday, and we probably have discussed it, I think.

We shouldn't be emitting an Info alert on every single block, as in most blocks the balances aren't expected to change. We should only be firing one when there's a balance change.

File: Challenge-3/src/helper.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    const l1BigNumber = BigNumber.from(l1Balance);
+    const l2BigNumber = BigNumber.from(totalSupply);
+
+    if (l1BigNumber.lt(l2BigNumber)) {
+      findings.push(
+        createFinding(l1Balance, l2BigNumber.toString(), l2Network),
+      );
+    }
+
+    return totalSupply.toString();
```
Comment (Vxatz) [inline Challenge-3/src/helper.ts:104]: This returned string is not really used anywhere

File: Challenge-3/src/agent.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    } else {
+      // For non-mainnet chains, fetch L2 supply details
+      await helperInstance.getL2Supply(
```
Comment (Vxatz) [inline Challenge-3/src/agent.ts:41]: The function doesn't only fetch the L2 supply, but also potentially creates a high severity finding. You could update both the comment and the method name to indicate this.

File: Challenge-3/src/findings.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  l2Supply: string,
+  network: string,
+): Finding => {
+  return Finding.fromObject({
+    name: `L1 Escrow has less balance than L2 supply on ${network}`,
+    description: `balance of ${l1Escrow}, ${network} l2Supply-> ${l2Supply}`,
+    alertId: "L2-SUPPLY",
+    severity: FindingSeverity.High,
+    type: FindingType.Exploit,
+    protocol: `${network}`,
```
Comment (Vxatz) [inline Challenge-3/src/findings.ts:14]: No need to use backticks on this

**Review by Vxatz — CHANGES_REQUESTED** [review]

File: Challenge-3/src/agent.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    const findings: Finding[] = [];
+    const helperInstance = new Helper(provider);
+
+    if (chainId === 1) {
+      // Concurrently fetch balances for specified escrow addresses
+      const [optBalance, abtBalance] = await Promise.all([
+        helperInstance.getL1Balance(OPT_ESCROW_ADDRESS, blockEvent.blockNumber),
+        helperInstance.getL1Balance(ABT_ESCROW_ADDRESS, blockEvent.blockNumber),
+      ]);
+      if (prevOptBalance !== optBalance || prevAbtBalance !== abtBalance || !prevOptBalance || !prevAbtBalance) {
```
Comment (Vxatz) [inline Challenge-3/src/agent.ts:39]:
```suggestion
      if (prevOptBalance !== optBalance || prevAbtBalance !== abtBalance) {
```

The "undefined" checks are redundant, as they would still be "not equal" to the last fetched balance.

File: Challenge-3/src/helper.ts (hunk 1 above)
Comment (Vxatz) [reply Challenge-3/src/helper.ts:60]: Unresolved the comment above, as the `alertId` is still incorrect.

**Review by Vxatz — CHANGES_REQUESTED** [review]

**Review by Vxatz — APPROVED** [review]
LGTM 🎉

**Pull request merged**